src/nsrt_learning/strips_learning/gen_to_spec_learner.py

```diff
@@ -214,6 +214,31 @@ def _backchain_one_pass(
                         tuple(var_to_obj[var] for var in pnad.op.parameters))
                     if len(param_opt_to_nec_pnads[option.parent]) == 0:
                         param_opt_to_nec_pnads[option.parent].append(pnad)
+                    segs_in_pnad = [
+                        datapoint[0] for datapoint in pnad.datastore
+                    ]
+                    # In this case, we want to move the segment from
+                    # another PNAD into the current PNAD. Note that
+                    # we don't have to recompute the PNAD's add
+                    # effects or preconditions because of the fact that
+                    # this PNAD was found by the _find_best_matching
+                    # function (which internally checks that the
+                    # preconditions and add effects are all correct).
+                    if segment not in segs_in_pnad:
+                        # Find PNAD that the segment is currently in.
+                        for seg_pnad in pnads_for_option:
+                            segs_in_seg_pnad = [
+                                datapoint[0]
+                                for datapoint in seg_pnad.datastore
+                            ]
+                            if segment in segs_in_seg_pnad:
+                                seg_idx = segs_in_seg_pnad.index(segment)
+                                seg_pnad.datastore.pop(seg_idx)
+                                break
+                        pnad.datastore.append((segment, var_to_obj))
+                        self._remove_empty_datastore_pnads(
```

Comment:
is this because the one you just removed this segment from may now be empty?

Reply:
exactly